### PR TITLE
Update Storm pivot docs

### PR DESCRIPTION
**Pull request by therealsilence**
- Revise / restructure / clarify pivot docs
- Update for more current examples / use cases
- Add / clarify examples for:
  - pivot / traverse to lists of forms
  - pivot / traverse to wildcard forms (partial form name matches)
  - pivot / traverse to interfaces
  - pivot to / from tags

### CODE REVIEW

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 1]
```diff
@@ -7,1027 +7,1721 @@
...
-    <inet:whois:rec> -> *
+In these cases, you can use **implicit pivot syntax** (implicit syntax) for your Storm query. Implicit
+syntax takes advantage of Synapse's :ref:`gloss-type-aware` to "know" which properties can be pivoted
+to (or from), given the source and target forms for the pivot operation. With implicit syntax, you do
+not need to specify the source or target property in cases where it is self-evident for the forms used.
 
+Using implicit syntax, we can rewrite the above query as follows:
 
-In the example above, the pivot would navigate from the ``:fqdn``, ``:registrar``,  and ``:registrant`` secondary properties of the ``inet:whois:rec`` nodes (for example) to the associated ``inet:fqdn``, ``inet:whois:rar``, and ``inet:whois:reg`` nodes.
+.. storm-pre:: inet:fqdn =  vertex.link -> inet:dns:a -> inet:ipv4
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:214]:
```suggestion
.. storm-pre:: inet:fqdn = vertex.link -> inet:dns:a -> inet:ipv4
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 2]
```diff
@@ -7,1027 +7,1721 @@
...
 ::
+  
+  <file:bytes nodes> -> it:host:activity:exe
 
-    <inet:dns:a> :ipv4 -+> inet:ipv4
+*Pivot from a set of usernames (inet:user nodes) to any file paths (file:path nodes) where the username is an element of the path (a :base value - for example, where a username might appear in a PDB path):*
 
+.. storm-pre:: [ inet:user=swanson file:path='c:\users\swanson\dev\projects\getthesnacks\release\getthesnacks.pdb' ]
+.. storm-pre:: inet:user=swanson -> file:path:base
+.. storm-pre:: inet:user=swanson -> file:path:base tree { -> file:path:dir }
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:393]: Getting fancy with the `tree` command, nice!

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 3]
```diff
@@ -7,1027 +7,1721 @@
...
 
+.. storm-pre:: inet:dns:a -> inet:fqdn
+.. storm-pre:: inet:dns:a :fqdn -> inet:fqdn
 
-.. storm-pre:: [media:news="*" +(refs)> { inet:fqdn=woot.com } ] -(refs)> inet:fqdn
 ::
+  
+  <inet:dns:a nodes> -> inet:fqdn
+
+The query above uses implicit syntax. Synapse recognizes that the, given a target form of ``inet:fqdn``,
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:488]:
```suggestion
The query above uses implicit syntax. Synapse recognizes that, given a target form of ``inet:fqdn``,
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 4]
```diff
@@ -7,1027 +7,1721 @@
...
+Secondary to Secondary Property Pivot
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-    <media:news> -(*)> *
+When pivoting from a secondary property of a set of source nodes to target nodes with the same secondary
+property (e.g., from the ``:ipv4`` property of a set of DNS A nodes to a set of network flow nodes with
+the same IPv4 as a ``:dst:ipv4`` property), the target can be:
 
+- a form name and property name; or
+- an interface name and property name.
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:555]: can also be a list of props like:
`inet:dns:a:ipv4=1.2.3.4 :ipv4 -> (inet:banner:server:ipv4, inet:flow:src:ipv4)`

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 5]
```diff
@@ -7,1027 +7,1721 @@
...
+  ::
+    
+    <inet:whois:rec nodes> -> inet:fqdn -> inet:dns:a
 
-**Examples:**
+*Pivot from a set of DNS requests (inet:dns:request nodes) to all host activity nodes (all nodes of all forms that inherit the it:host:activity interface) that share the same file as their :exe property:*
 
-*Traverse the "refs" light edge from an article and join the article with the FQDNs "referenced" by the article:*
+.. storm-pre:: [ file:bytes=sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855 (inet:dns:request=* :query:name:fqdn=woot.com :exe=sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) (it:exec:file:write=* :exe=sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855 :path='c:/windows/scvhost.exe') (it:exec:file:read=* :exe=sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855 :path='c:/windows/autoexec.bat') ]
+.. storm-pre:: inet:dns:request:exe -> it:host:activity:exe
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:589]:
```suggestion
.. storm-pre:: inet:dns:request :exe -> it:host:activity:exe
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 6]
```diff
@@ -7,1027 +7,1721 @@
...
+- Pivot from a set of FQDNs to the DNS A records that reference them:
+  
+  ``<inet:fqdn nodes> <- inet:dns:a``
 
-*Join a set of IPv4 address with the set of any nodes linked to those IPv4 addresses via the "hasip" or "ipwhois" light edges:*
+Because Storm evaluates operations from left to right, the "pivot out" arrow is generally more intuitive and
+has been used to implement nearly all pivot operations in Storm (the second example, above, is not supported
+and will generate a ``StormRuntimeError``).
+
+The pivot in operation is a specialized operation that can only be used with the wildcard ( ``*`` ) as a
```
[the remark this answers is not in the document]
Comment (Cisphyx) [reply docs/synapse/userguides/storm_ref_pivot.rstorm:619]: Nevermind, I see that this is covered separately later on.

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 7]
```diff
@@ -7,1027 +7,1721 @@
...
-- The Storm :ref:`storm-tee` command can be used to perform concurrent pivot in and walk / pivot out and walk operations on an inbound set of nodes:
-  
-.. storm-pre:: inet:dns:a inet:fqdn | tee { --> * } { <-- * }
 ::
   
-  <query> | tee { --> * } { <-- * }
+  inet:http:request:header:name = referer $url = :value -> { inet:url ?= $url }
 
-.. _pivot-out-and-join:
+The query above illustratres using raw pivot syntax to pivot from a secondary to a primary property where the
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:712]:
```suggestion
The query above illustrates using raw pivot syntax to pivot from a secondary to a primary property where the
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 8]
```diff
@@ -7,1027 +7,1721 @@
...
+
+*<query>* *<pivot_operator>* **#** *<string>* | ***** | ****** [ **.** *<string>*  | ***** | ****** ... ]
 
-**Usage Notes:**
 
-- The pivot in and walk operator can only be used with a wildcard ( ``*`` ); it is not possible to specify a particular form as the target of the operation. A filter operation can be used to refine the results of the pivot and walk operation if necessary.
+*Pivot from a set of nodes to the tags (syn:tag nodes) associated with any third-party reporting where the third tag element is "bisonal":*
 
-- The pivot and walk operators (pivot out and walk / pivot in and walk) are useful for "exploring" data in a Cortex as they will return all the nodes "next to" the working set of nodes (subject to :ref:`gloss-type-aware`) without requiring the user to have specific knowledge of the data model.
+.. storm-pre:: [ file:bytes=sha256:89f250599e09f8631040e73cd9ea5e515d87e3d1d989f484686893becec1a9bc +#rep.alienvault.bisonal +#rep.malwarebazaar.bisonal +#rep.malwarebazaar.3p.intetzer.bisonal ]
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:843]:
```suggestion
.. storm-pre:: [ file:bytes=sha256:89f250599e09f8631040e73cd9ea5e515d87e3d1d989f484686893becec1a9bc +#rep.alienvault.bisonal +#rep.malwarebazaar.bisonal +#rep.malwarebazaar.3p.intezer.bisonal ]
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 9]
```diff
@@ -7,1027 +7,1721 @@
...
+You can also use the pivot in operator to pivot from the ``:n1`` property by default:
 
+::
+  
+  <edge:has nodes> <- *
 
-*Pivot to ALL tag nodes:*
 
-- Pivot from a set of files to the ``syn:tag`` nodes for **all** tags applied to those files (i.e., each tag in each tag tree applied to each node, from root to leaf):
+*Pivot from a set of "has" edge nodes to any vulnerabilities taht the objects "have":*
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:1195]:
```suggestion
*Pivot from a set of "has" edge nodes to any vulnerabilities that the objects "have":*
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 10]
```diff
@@ -7,1027 +7,1721 @@
...
+Traversal operations are used to navigate between sets of nodes that are linked using a lightweight (light)
+edge. Traversal is still considered a "pivot" between sets of nodes, but is named differently to distinguish
+light edge navigation from property-to-property pivot navigation.
 
-.. _pivot-from-tags:
+.. TIP::
+  
+  There are no light edges (i.e., specific light edge verbs) defined within Synapse by default. Users can create
+  and define their own according to their needs. However, both the Synapse Data Model :ref:`dm-forms` documentation
+  and the Optic UI's `Data Model Explorer`_ list edges used by The Vertex Project by convention and which are
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:1262]:
```suggestion
  and the Optic UI's `Data Model Explorer`_ list edges used by The Vertex Project by convention which are
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 11]
```diff
@@ -7,1027 +7,1721 @@
...
-- Pivot to all domains tagged with tags from any of the inbound ``syn:tag`` nodes:
+- ``-(<verb>)>``
 
+``<verb>`` can be a single edge, a list of edges, or a wildcard.
 
-.. storm-pre:: syn:tag -> inet:fqdn
-::
+Unlike property-to-property relationships, edge relationships have a **direction**. There is a "source" node
+(``n1``) and a "target" node (``n2``) for the relationship itself; an article (``media:news`` node, the ``n1``)
+can refeference an indicator such as a hash (``hash:md5`` node, the ``n2``) but it does not make sense for a
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:1286]:
```suggestion
can reference an indicator such as a hash (``hash:md5`` node, the ``n2``) but it does not make sense for a
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 12]
```diff
@@ -7,1027 +7,1721 @@
...
+- a partial form name (wildcard match, e.g., ``hash:*``);
+- an interface name (e.g., ``it:host:activity``);
+- a list of form names (e.g., ``( hash:sha256, file:bytes )``); or
+- a wildcard / asterisk ( ``*`` ).
 
+.. NOTE::
+  
+  You cannot specify property **values** in traversal operations. For example, the following is invalid:
+  
+  ``media:news -(refs)> inet:fqdn=woot.com``
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:1314]: We do actually allow a comparator on walk targets, you can do things like:
```
> inet:fqdn=foo.com -(refs)> inet:fqdn^=bar
inet:fqdn=bar.com
	.created = 2021/01/29 15:39:20.398
```

Comment (Cisphyx) [reply docs/synapse/userguides/storm_ref_pivot.rstorm:1314]: There's a story in the backlog for documenting that, SYN-2427

Comment (Cisphyx) [reply docs/synapse/userguides/storm_ref_pivot.rstorm:1314]: Also just FYI, you can currently target secondary props and use cmprs on them in walks, but that is being removed in 3.0.

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 13]
```diff
@@ -7,1027 +7,1721 @@
...
----------------------
+.. storm-pre:: [media:news="*" +(refs)> { inet:fqdn=woot.com } ]
+.. storm-pre:: media:news -(refs)> *
 
-Pivot operations in Storm can always be executed by **explicitly** specifying the source and target properties for the pivot. This is referred to as **explicit pivot syntax** or explicit syntax. For example:
+::
+  
+  <media:news> -(refs)> *
+
+*Traverse the "ipwhois" light edge from a set of IPv4 addresses (inet:ipv4 nodes) to the network registration / network whois records (inet:whois:iprec nodes) the IPs are associated with:*
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:1355]:
```suggestion
*Traverse the "ipwhois" light edge from a set of IPv4 addresses (inet:ipv4 nodes) to the network registration / network WHOIS records (inet:whois:iprec nodes) the IPs are associated with:*
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 14]
```diff
@@ -7,1027 +7,1721 @@
...
+.. storm-pre:: [ (inet:whois:iprec=* :asof=now :name=LVLT-GOGL-8-8-8 :net4:min=8.8.8.0 :net4:max=8.8.8.255) ] [ +(ipwhois)> { [ inet:ipv4=8.8.8.8 ] } ]
+.. storm-pre:: inet:ipv4 <(ipwhois)- inet:whois:iprec
 
-.. storm-pre:: inet:fqdn=vertex.link -> inet:dns:a:fqdn :ipv4 -> inet:ipv4
 ::
   
-  inet:fqdn=vertex.link -> inet:dns:a:fqdn :ipv4 -> inet:ipv4
+  <inet:ipv4 nodes> <(ipwhois)- inet:whois:iprec
+  
+*Traverse the "seen" light edges from set of DNS A records (inet:dns:a nodes) to the sources (meta:source nodes) that "saw" (observed or provided data for) the record:*
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:1364]:
```suggestion
*Traverse the "seen" light edges from a set of DNS A records (inet:dns:a nodes) to the sources (meta:source nodes) that "saw" (observed or provided data for) the record:*
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 15]
```diff
@@ -7,1027 +7,1721 @@
...
 
-.. storm-pre:: inet:dns:a :ipv4 -> inet:ipv4
 ::
+  
+  risk:vuln <(*)- *
 
-    <inet:dns:a> :ipv4 -> inet:ipv4
+Vertex Project conventions can link ``risk:vuln`` nodes to things that have the vulnerability ( via ``<(has)-``
+edges) or things that use the vulnerability ( via ``<(uses)-`` edges - e.g., a threat cluster (``risk:threat``),
+a file (``file:bytes``) that makes use of an exploit for the vulnerability, etc.
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:1448]: Missing a closing paren that I think should be at the end of the line:
```suggestion
Vertex Project conventions can link ``risk:vuln`` nodes to things that have the vulnerability ( via ``<(has)-``
edges) or things that use the vulnerability ( via ``<(uses)-`` edges - e.g., a threat cluster (``risk:threat``),
a file (``file:bytes``) that makes use of an exploit for the vulnerability, etc.)
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 16]
```diff
@@ -7,1027 +7,1721 @@
...
-**Implicit syntax:**
+.. _storm-pivot-traverse:
 
+Pivot and Traverse Operations
+-----------------------------
 
-.. storm-pre:: inet:dns:a -> inet:ipv4
-::
+Property pivots and edge traversals can be combined into a single "pivot and traverse" operation to perform "both types"
+of navigation simultaneoulsy from a set of source nodes. "Pivot and traverse" combines a pivot operation with a traverse
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:1457]:
```suggestion
of navigation simultaneously from a set of source nodes. "Pivot and traverse" combines a pivot operation with a traverse
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 17]
```diff
@@ -7,1027 +7,1721 @@
...
-*Implicit syntax with multiple source properties:*
+.. _pivot-join:
 
-- Pivot from a set of files to their associated SHA256 hashes:
+Pivot and Join
+++++++++++++++
 
-**Implicit syntax:**
+The pivot and join operator ( ``-+>`` ) is the pivot "arrow" with an embedded plus sign ( ``+`` ) to represent joining
+the source and target nodes. The operator can be used in anywhere the standard pivot operator ( ``->`` ) is used, including
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:1561]:
```suggestion
the source and target nodes. The operator can be used anywhere the standard pivot operator ( ``->`` ) is used, including
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 18]
```diff
@@ -7,1027 +7,1721 @@
...
+  <file:bytes nodes> <+- *
+
+
+.. _join-light-edge:
+
+Traverse and Join
++++++++++++++++++
+
+The traverse and join operator ( ``-(<verb>)+>`` ) is a traversal "arrow" with a plus sign ( ``+`` ) to represent
+joining the source and target nodes. The operator can be used in anywhere the standard ( ``-(<verb>)>`` ) traversal
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:1631]:
```suggestion
joining the source and target nodes. The operator can be used anywhere the standard ( ``-(<verb>)>`` ) traversal
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm [hunk 19]
```diff
@@ -7,1027 +7,1721 @@
...
+
+*Join a threat cluster (risk:threat node) with any nodes used or targeted by the cluster:*
+
+.. storm-pre:: [ ( risk:threat=* :org:name='sparkling unicorn' ) (risk:vuln=* :cve=cve-2012-0158) ]
+.. storm-pre:: risk:threat:org:name='sparkling unicorn' [ +(uses)> { risk:vuln:cve=cve-2012-0158 } ] [ +(targets)> { ou:org:alias=vertex } ]
+.. storm-pre:: risk:threat -( (uses, targets) )+> *
+
+::
+  
+  <risk:threat> -( (uses, targets) )> *
```
Comment (Cisphyx) [inline docs/synapse/userguides/storm_ref_pivot.rstorm:1673]:
```suggestion
  <risk:threat> -( (uses, targets) )+> *
```

File: docs/synapse/userguides/storm_ref_pivot.rstorm (hunk 15 above)
Comment (therealsilence) [reply docs/synapse/userguides/storm_ref_pivot.rstorm:1448]: Cleaned up

File: docs/synapse/userguides/storm_ref_pivot.rstorm (hunk 12 above)
Comment (therealsilence) [reply docs/synapse/userguides/storm_ref_pivot.rstorm]: Updated

File: docs/synapse/userguides/storm_ref_pivot.rstorm (hunk 4 above)
Comment (therealsilence) [reply docs/synapse/userguides/storm_ref_pivot.rstorm:556]: Updated

**Review by Cisphyx — APPROVED** [review]

**Pull request merged**